source/extensions/filters/network/local_ratelimit/local_ratelimit.cc

```diff
@@ -0,0 +1,74 @@
+#include "extensions/filters/network/local_ratelimit/local_ratelimit.h"
+
+#include "envoy/event/dispatcher.h"
+
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace LocalRateLimitFilter {
+
+Config::Config(
+    const envoy::config::filter::network::local_rate_limit::v2alpha::LocalRateLimit& proto_config,
+    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime)
+    : fill_timer_(dispatcher.createTimer([this] { onFillTimer(); })),
+      max_tokens_(proto_config.token_bucket().max_tokens()),
+      tokens_per_fill_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.token_bucket(), tokens_per_fill, 1)),
+      fill_interval_(PROTOBUF_GET_MS_REQUIRED(proto_config.token_bucket(), fill_interval)),
+      enabled_(proto_config.enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)), tokens_(max_tokens_) {
+  fill_timer_->enableTimer(fill_interval_);
+}
+
+LocalRateLimitStats Config::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  const std::string final_prefix = "local_rate_limit." + prefix;
+  return {ALL_LOCAL_RATE_LIMIT_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
+}
+
+void Config::onFillTimer() {
+  uint32_t new_tokens_value;
+  {
+    // TODO(mattklein123): Consider replacing with an atomic CAS loop.
+    absl::MutexLock lock(&mutex_);
```

Comment:
You might be able to drop the mutex and go with best-effort enforcement here and in `canCreateConnection()`. Since it's just `tokens_` that you're guarding with the mutex, consider incrementing an atomic counter in `onNewConnection()` rather than decrementing tokens. Then in `onFillTimer()` you just subtract from the counter `tokens_per_fill_`. This would allow you to just check against some upper bound and the logic would be almost identical to what's happening in the gradient controller for adaptive concurrency:
https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc#L178

Reply:
Interesting idea. Still, you would need a CAS loop in the fill timer callback to avoid underflow. I also think that we should be exact on the rate limit.

TBH the CAS implementation is simple, I can just do it now. I didn't because I thought people would complain about premature optimization.